internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,332 @@
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memDB.Dirty() || p.len > 0
+}
+
+// GetMemDB implements MemBuffer interface.
+func (p *PipelinedMemDB) GetMemDB() *MemDB {
+	panic("GetMemDB should not be invoked for PipelinedMemDB")
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemDB > flushed memdbs.
+func (p *PipelinedMemDB) Get(ctx context.Context, k []byte) ([]byte, error) {
+	v, err := p.memDB.Get(k)
+	if err == nil {
+		return v, nil
+	}
+	if !tikverr.IsErrNotFound(err) {
+		return nil, err
+	}
+	if p.flushingMemDB != nil {
+		v, err = p.flushingMemDB.Get(k)
+		if err == nil {
+			return v, nil
+		}
+		if !tikverr.IsErrNotFound(err) {
+			return nil, err
+		}
+	}
+	if ctx.Value(pipelinedMemDBSkipRemoteBufferKey) != nil {
+		return nil, tikverr.ErrNotExist
+	}
+	// read remote buffer
+	var (
+		dataMap map[string][]byte
+		ok      bool
+	)
+	dataMap, err = p.bufferBatchGetter(ctx, [][]byte{k})
+	if err != nil {
+		return nil, err
+	}
+	v, ok = dataMap[string(k)]
+	if !ok {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memDB.GetFlags(k)
+	if p.flushingMemDB != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemDB.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+func (p *PipelinedMemDB) UpdateFlags(k []byte, ops ...kv.FlagsOp) {
+	p.memDB.UpdateFlags(k, ops...)
+}
+
+// Set sets the value for key k in the MemBuffer.
+func (p *PipelinedMemDB) Set(key, value []byte) error {
+	return p.memDB.Set(key, value)
+}
+
+// SetWithFlags sets the value for key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) SetWithFlags(key, value []byte, ops ...kv.FlagsOp) error {
+	return p.memDB.SetWithFlags(key, value, ops...)
+}
+
+// Delete deletes the key k in the MemBuffer.
+func (p *PipelinedMemDB) Delete(key []byte) error {
+	return p.memDB.Delete(key)
+}
+
+// DeleteWithFlags deletes the key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) DeleteWithFlags(key []byte, ops ...kv.FlagsOp) error {
+	return p.memDB.DeleteWithFlags(key, ops...)
+}
+
+// Flush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemDB is done.
+// The first returned value indicates whether the flush is triggered.
+// The second returned value is the error if there is a failure, txn should abort when there is an error.
+// When the mutable memdb is too large, it blocks until the ongoing flush is done.
+func (p *PipelinedMemDB) Flush(force bool) (bool, error) {
+	if p.flushFunc == nil {
+		return false, errors.New("flushFunc is not provided")
+	}
+	if !force && !p.needFlush() {
+		return false, nil
+	}
+	if p.flushingMemDB != nil {
+		if err := <-p.errCh; err != nil {
+			return false, err
```

Comment:
We may need to set p.flushingMemDB to nil before return, otherwise, `Flush` (and also `FlushWait`) must NOT be called once an error returned. Think about the following case:
- let p = { flushingMemDB: nil }
- p.Flush => p.flushingMemDB = x; go flush() { p.errCh <- e };
- p.Flush => return e (since p.flushingMemDB != nil and e in p.errCh)
- p.Flush => block forever (since p.flushingMemDB != nil but nothing in p.errCh)

Reply:
Since `p.Flush()` already return error, the transaction should be aborted and the final `p.Flush()` is unreachable.

May need to store the error status and return error directly for the `Flush` calls after first error is produced.